py7zr/compression.py

```diff
@@ -293,34 +304,42 @@ def extract_single(self, fp: BinaryIO, files, src_start: int) -> None:
             if f.emptystream:
                 fileish.write(b'')
             else:
-                self.decompress(fp, f.folder, fileish, f.uncompressed[-1], f.compressed)
+                self.decompress(fp, f.folder, fileish, f.uncompressed[-1], f.compressed, src_end)
             fileish.close()
 
     def decompress(self, fp: BinaryIO, folder, fileish: Handler,
-                   size: int, compressed_size: Optional[int]) -> None:
-        """decompressor wrapper called from extract method."""
+                   size: int, compressed_size: Optional[int], src_end: int) -> None:
+        """decompressor wrapper called from extract method.
+
+           :parameter fp: archive source file pointer
+           :parameter folder: Folder object that have decompressor object.
+           :parameter fileish: output file Handler, BufferHandler, FileHandler or NullHandler
+           :parameter size: uncompressed size of target file.
+           :parameter compressed_size: compressed size of target file.
+           :parameter src_end: end position of the folder
+           :returns None
+        """
         assert folder is not None
         out_remaining = size
         decompressor = folder.get_decompressor(compressed_size)
         while out_remaining > 0:
-            if decompressor.eof:
-                raise DecompressionError
             max_length = min(out_remaining, io.DEFAULT_BUFFER_SIZE)
-            if decompressor.needs_input:
-                read_size = min(Configuration.get('read_blocksize'), decompressor.remaining_size)
+            rest_size = src_end - fp.tell()
```

Comment:
It should not use 'src_end' when non-solid archives with non-multi-thread extraction.
it is better to retrieve folder size from 'compressed_size'